### PR TITLE
Heretic Ascension Tweaks

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Some changes to Cosmic and Knock heretic ascensions, in response to feedback.

- Cosmic Heretics can no longer control their summon while jaunted. 
- Additionally the death link element the Star Gazer used... wouldn't work if there was more than one cosmic heretic. I refactored it into a component, so that it would.

Frankly there are probably a lot of _other_ abilities which shouldn't be usable while jaunted but are, but I have been burned in the past by adding defaults which were applied too widely so we'll leave it case-by-case for now.

- The Knock rift can no longer summon Flesh Worms of any kind.
- But it _can_ summon fire sharks (we really need to reflavour these...) and any future "basic mob" heretic mobs (or ones which are converted).
- Additionally the rift can't be clicked by ghosts while it's polling ghosts automatically because that would create funky empty-minded mobs.
- Finally it goes away when the heretic dies.

## Why It's Good For The Game

For Cosmic Heretics, controlling the Star Gazer in conjunction with Space Jaunt essentially meant that the heretic was playing an RTS as an invincible observer with an invulnerable unit, now they have to actually be present and killable in order to sic The Beast on you.

For Knock Heretics, Minor Flesh Worms are a proof of concept which was never meant to be used except by admins. They're _barely_ weaker than a normal flesh worm, extraordinarily tanky, delete walls, and generally devalue the Flesh path ascension.
Vanishing upon heretic death is because every _other_ ascension is at least theoretically stoppable except this one. Now this one is too.
It's still _very hard_ to kill the heretic because every _10 seconds_ they can transform into a heretic mob which acts as an extra health pool.

## Changelog

:cl:
fix: If two cosmic heretics ascend in the same round, their star gazer survival will be linked to each individual heretic and not shared by just one of them.
fix: You can't click the Knock heretic portal to join as a mob while already signed up to become a mob.
balance: Cosmic heretics can't order the Star Gazer around while jaunting.
balance: The Knock Heretic portal cannot summon Flesh Worms, but can summon Fire Sharks.
balance: The Knock Heretic portal will disperse if its creator is killed.
/:cl:

### CODE REVIEW

**Comment by dieamond13** [comment]
The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO

**Comment by AtomTheProphet** [comment]
> The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO

Yeah, absolutely. This just feels like an unnecessarily heavy nerf to a path that has relatively weak combat potential compared to other paths.

**Comment by Jacquerel** [comment]
> The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO

once it ascends it has both of these things via constantly transforming into mobs which give it extra repleneshing health, strong attacks and damaging abilities, and are immune to stamina damage and many forms of paralysis

**Comment by AtomTheProphet** [comment]
> > The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO
> 
> once it ascends it has both of these things via constantly transforming into mobs which give it extra repleneshing health, strong attacks and damaging abilities, and are immune to stamina damage and many forms of paralysis

Maybe health should somehow transfer to what they're transforming into then?

**Comment by Jacquerel** [comment]
> > > The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO
> > 
> > 
> > once it ascends it has both of these things via constantly transforming into mobs which give it extra repleneshing health, strong attacks and damaging abilities, and are immune to stamina damage and many forms of paralysis
> 
> Maybe health should somehow transfer to what they're transforming into then?

that would be a very simple change and a far more enormous reduction in strength than anything in this PR

**Comment by Elli-Skala** [comment]
> essentially meant that the heretic was playing an RTS as an invincible observer 

But what if....

File: code/datums/actions/action.dm [hunk 1]
```diff
@@ -103,6 +103,8 @@
 		RegisterSignal(owner, SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), PROC_REF(update_status_on_signal))
 	if(check_flags & AB_CHECK_LYING)
 		RegisterSignal(owner, COMSIG_LIVING_SET_BODY_POSITION, PROC_REF(update_status_on_signal))
+	if(check_flags & AB_CHECK_PHASED)
+		RegisterSignal(owner, SIGNAL_ADDTRAIT(TRAIT_MAGICALLY_PHASED), PROC_REF(update_status_on_signal))
```
Comment (MrMelbert) [inline code/datums/actions/action.dm:107]: Also register `SIGNAL_REMOVETRAIT`

File: code/__DEFINES/actions.dm [hunk 2]
```diff
@@ -8,13 +8,16 @@
 #define AB_CHECK_CONSCIOUS (1<<3)
 ///Action button checks if user is incapacitated
 #define AB_CHECK_INCAPACITATED (1<<4)
+///Action button checks if user is jaunting
+#define AB_CHECK_PHASED (1<<5)
```
Comment (MrMelbert) [inline code/__DEFINES/actions.dm:12]: I almost feel like this should replace `SPELL_CASTABLE_WHILE_PHASED` req on spell actions

**Comment by AtomTheProphet** [comment]
> > The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO
> 
> once it ascends it has both of these things via constantly transforming into mobs which give it extra repleneshing health, strong attacks and damaging abilities, and are immune to stamina damage and many forms of paralysis

Hey, I don't quite think the shapechange is able to heal someone (I tested it last night). Am I missing something?

**Comment by Jacquerel** [comment]
> > > The portal going away right when the heretic dies is a bit much for a heretic ascension that doesnt get damage or stun resistance, even voids AOE damage is still there when they die. Too hard of a nerf because of that IMO
> > 
> > 
> > once it ascends it has both of these things via constantly transforming into mobs which give it extra repleneshing health, strong attacks and damaging abilities, and are immune to stamina damage and many forms of paralysis
> 
> Hey, I don't quite think the shapechange is able to heal someone (I tested it last night). Am I missing something?

It does not heal you, but any damage you take while shapeshifted does not transfer to you. Effectively you can gain 150 additional health points by turning into a Flesh Stalker once every ten seconds.

**Comment by mc-oofert** [comment]
does shapeshifting preserve the signal tomfoolery on the knock ascension portal

**Comment by Jacquerel** [comment]
> does shapeshifting preserve the signal tomfoolery on the knock ascension portal

Yes, nothing happens to the portal until you die as a human.
Also I forgot that knock transformations actually have even more health than usual, so the Stalker gives you more HP than it takes to completely kill a human (from health > crit > hard crit > dead).

**Comment by MrrFiish** [comment]
I honestly don't even like the portal getting a cooldown per person, if you let a heretic ascend it should be OVER, I imagine a flood of heretic minions coming from the portal and if it is like 3 dude with 3 minute cooldowns it won't do much at all.

**Comment by Jacquerel** [comment]
fine I'll remove the cooldown

**Comment by MrrFiish** [comment]
> fine I'll remove the cooldown

You are SO cool thank you for considering it :D

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**